source/simulator/helper_functions.cc

```diff
@@ -1698,64 +1736,43 @@ namespace aspect
-                                                                   /*dof index within component=*/ j);
+            {
+              const unsigned int temperature_idx
+                = dof_handler.get_fe().component_to_system_index(introspection.component_indices.temperature,
+                                                                 /*dof index within component=*/ j);
 
-                // skip entries that are not locally owned:
-                if (dof_handler.locally_owned_dofs().is_element(local_dof_indices[temperature_idx]))
-                  {
+              // skip entries that are not locally owned:
+              if (dof_handler.locally_owned_dofs().is_element(local_dof_indices[temperature_idx]))
+                {
+                  if (temperature_and_composition_use_same_fe)
+                    distributed_vector(local_dof_indices[temperature_idx]) = in_C.temperature[j];
+                  else
                     distributed_vector(local_dof_indices[temperature_idx]) = in_T.temperature[j];
-                    distributed_reaction_vector(local_dof_indices[temperature_idx]) = accumulated_reactions_T[j];
-                  }
-              }
+
+                  distributed_reaction_vector(local_dof_indices[temperature_idx]) = accumulated_reactions_T[j];
+                }
+            }
         }
 
+    distributed_vector.compress(VectorOperation::insert);
+    distributed_reaction_vector.compress(VectorOperation::insert);
+
     // put the final values into the solution vector
     for (unsigned int c=0; c<introspection.n_compositional_fields; ++c)
       {
         const unsigned int block_c = introspection.block_indices.compositional_fields[c];
-        distributed_vector.block(block_c).compress(VectorOperation::insert);
-        solution.block(block_c) = distributed_vector.block(block_c);
-
-        // we have to update the old solution with our reaction update too
-        // so that the advection scheme will have the correct time stepping in the next step
-        distributed_reaction_vector.block(block_c).compress(VectorOperation::insert);
-
-        // we do not need distributed_vector any more, use it to temporarily store the update
-        distributed_vector.block(block_c) = old_solution.block(block_c);
-        distributed_vector.block(block_c) +=  distributed_reaction_vector.block(block_c);
-        old_solution.block(block_c) = distributed_vector.block(block_c);
-
-        distributed_vector.block(block_c) = old_old_solution.block(block_c);
-        distributed_vector.block(block_c) +=  distributed_reaction_vector.block(block_c);
-        old_old_solution.block(block_c) = distributed_vector.block(block_c);
-
-        operator_split_reaction_vector.block(block_c) = distributed_reaction_vector.block(block_c);
+        update_solution_vectors_with_reaction_results(block_c,distributed_vector,distributed_reaction_vector);
```

Comment:
just inline the initializer of `block_c` into the call here and get rid of the curly braces